source/common/stats/symbol_table_impl.cc

```diff
@@ -117,76 +144,124 @@ SymbolEncoding SymbolTable::encode(const absl::string_view name) {
-  // If the string segment doesn't already exist,
-  if (encode_find == encode_map_.end()) {
-    // We create the actual string, place it in the decode_map_, and then insert a string_view
-    // pointing to it in the encode_map_. This allows us to only store the string once.
-    std::string str = std::string(sv);
+Symbol SymbolTable::toSymbol(absl::string_view sv) {
+  {
+    // First try to find the symbol with just a read-lock, so concurrent
+    // lookups for an already-allocated symbol do not contend.
+    absl::ReaderMutexLock lock(&lock_);
+    auto encode_find = encode_map_.find(sv);
+    if (encode_find != encode_map_.end()) {
+      // Increment the refcount of the already existing symbol. Note that the
+      // ref_count_ is atomic to allow incrementing it under read-lock.
+      SharedSymbol& shared_symbol = *encode_find->second;
+      ++(shared_symbol.ref_count_);
+      return shared_symbol.symbol_;
+    }
+  }
 
+  // If the find() under read-lock failed, we need to release it and take a
+  // write-lock. Note that another thread may race to insert the symbol during
+  // this window of time with the lock released, so we need to check again
+  // under write-lock, which we do by proactively allocating the string and
+  // attempting to insert it into the encode_map. If that worked, we also
+  // write the decode-map, transferring the ownership of the string to the
+  // decode-map value.
+  absl::MutexLock lock(&lock_);
+
+  // If the string segment doesn't already exist, create the actual string as a
+  // nul-terminated char-array and insert into encode_map_, and then insert a
+  // string_view pointing to it in the encode_map_. This allows us to only store
+  // the string once.
+  size_t size = sv.size() + 1;
+  std::unique_ptr<char[]> str = std::make_unique<char[]>(size);
+  StringUtil::strlcpy(str.get(), sv.data(), size);
+
+  auto encode_insert =
+      encode_map_.insert({str.get(), std::make_unique<SharedSymbol>(next_symbol_)});
+  SharedSymbol& shared_symbol = *encode_insert.first->second;
+
+  if (encode_insert.second) {
+    // The insertion took place.
     auto decode_insert = decode_map_.insert({next_symbol_, std::move(str)});
     ASSERT(decode_insert.second);
-
-    auto encode_insert = encode_map_.insert({decode_insert.first->second, {next_symbol_, 1}});
-    ASSERT(encode_insert.second);
-
-    result = next_symbol_;
     newSymbol();
   } else {
-    // If the insertion didn't take place, return the actual value at that location and up the
-    // refcount at that location
-    result = encode_find->second.symbol_;
-    ++(encode_find->second.ref_count_);
+    // If the insertion didn't take place -- due to another thread racing to
+    // insert the same symbmol after we drop the read-lock above -- we can
+    // return the shared symbol, but we must bump the refcount.
+    ++(shared_symbol.ref_count_);
+
+    // Note: this condition is hard to hit in tests as it requires a tight race
```

Comment:
I see you also going to a lot of hoops to test various interleavings in the test code also via a custom notification class. if you feel it's very important to test specific interleavings, did you consider add a test hook concept to the code? Depending on how you do this I think the hook calls can probably be optimized out entirely and only hit during test code at which point you can do whatever you want to various threads? WDYT?

Reply:
Yes this occurred to me; do you have a pattern where this would be optimized out entirely in production, other than ifdef?  ifdef is probably OK though; just wanted to see if you have a trick.

Reply:
ifdef would definitely work, but, naively, I suspect that if you have a TestHook class with virtual functions, and as long as the null implementation is in the same compilation unit, I think it might be completely de-virtualized and compiled out, but I'm not sure.

Reply:
I see how the null implementation could be in the same compilation unit, but I don't see how the compiler could be sure it would be used if we have the same .o being tested as what we have in production.  

I decided to do it via a settable function hook, which could presumably be branch-predicted away in production. I'm also going to leave in the existing race test and the custom version of Notify because it gets us more contention.